### PR TITLE
Video Phone

### PR DESCRIPTION
**Pull request by Kurfursten**
> It's me. I have an operational update on Space Station 13. What? Can it wait until after my report?
> ... No. You're telling me the Organization has infiltrated farther than we thought? Now they have agents controlling both Nanotrasen AND the Syndicate? You mean we'll need to purchase our burner phones from the traders instead? If it is the will of Stein's Gate...

image

The Video Phone comes with 7 ringtones (set with ctrl click) and a vibrate setting (toggle with alt click)
Once you call someone, they have 8 seconds to pick up, or they receive a missed call from your number
The video phone will hang up if put somewhere its camera is obscured, like a bag (TODO: pockets, too)
It allows you to talk to someone in an alternate means from radio, and also see them (and a 1 tile radius around them).

**Video phones aren't on the telecommunications network, so they make great "off the grid" communication, and the ability to send an image might be useful in showing off hostages or bombs. Or showing off your bar project/new outfit if you're less traitorous.** The flipside of being off telecoms is that these ancient means of communication are totally incompatible with borgs and AI.

KNOWN ISSUES/TODO
* You can click things on the phone screen. This isn't an issue for most things since we have an examine distance cap but it lets you aim guns through the phone to fire at great distance or figure out which direction the other person is in. That's bad.
* Hang up when in pocket
* Testing
* Compatibility with non-video phones?

🆑 
* rscadd: Traders are now selling cheap burner phones for those paranoid about control by the organization. They have a video function built in!

### CODE REVIEW

**Comment by anunknownentity03** [comment]
give us a demo whenever yer able.

**Comment by Kharkov72** [comment]
He bought?

**Comment by SonixApache** [comment]
polaris had something like these, they're pretty cool
very cute addition kufrost, i hope you can finish it

File: code/game/objects/items/trader.dm [hunk 1]
```diff
@@ -1038,4 +1038,221 @@
...
+	if(video_display)
+		video_display.vis_contents.Cut()
+		video_display.vis_contents = block_radius(src, PHONE_VIEW)
+
+/obj/item/device/videophone/dropped(mob/user)
+	user.client.close_popup("phone[number]")
+	if(!istype(loc,/turf) && !istype(loc,/mob/living/carbon))
+		hangup()
+
+/obj/item/device/videophone/Hear(var/datum/speech/speech, var/rendered_speech="")
```
Comment (Kurfursten) [inline code/game/objects/items/trader.dm:1263]: Add a check here so it can only hear the person holding it.

**Comment by ShiftyRail** [comment]
1. Ree trader creep !

2. 
>but it lets you aim guns through the phone to fire at great distance or figure out which direction the other person is in

100% a feature : https://www.youtube.com/watch?v=pzDgaGTARqE
(maybe keep it as a traitor-only thing?)

3. Could you try to plug in the red phones we have to the network ? Don't do much, but would be pretty good for flavour.

File: code/game/objects/items/trader.dm [hunk 2]
```diff
@@ -1038,4 +1038,221 @@
...
+				connection = VP
+				status = PHONE_ONCALL //we're "already on the call"
+				event_key = user.on_moved.Add(src, "mob_moved") //begin tracking our movement
+				playsound(src, 'sound/items/phone/calling.ogg', 50, FALSE)
+			return
+	to_chat(user,"<span class='warning'>The line wasn't in use...</span>")
+	playsound(src, 'sound/items/phone/busy.ogg', 50, FALSE)
+
+/obj/item/device/videophone/proc/begin_ring(var/obj/item/device/videophone/VP)
+	if(!(status == PHONE_READY))
```
Comment (ShiftyRail) [inline code/game/objects/items/trader.dm]: `status != PHONE_READY` ?

File: code/game/objects/items/trader.dm [hunk 3]
```diff
@@ -1038,4 +1038,221 @@
...
+		//Create a view for the caller
+		video_display = Cb.setup_popup("phone[VP.number]",3,3,2,type=/obj/abstract/screen/noclick)
+		refresh_view()
+
+	var/client/Ca = user.client
+	//Create our view of the caller
+	VP.video_display = Ca.setup_popup("phone[number]",3,3,2,type=/obj/abstract/screen/noclick)
+	VP.refresh_view()
+
+proc/block_radius(var/atom/center, var/radius)
```
Comment (ShiftyRail) [inline code/game/objects/items/trader.dm]: please add a slash before the `proc` it makes me sad

**Comment by Kurfursten** [comment]
@ShiftyRail Unfortunately getting red phones involved would most likely involve changing their path which means it would become a MAPPING PR

Perhaps in a followup PR that would be small in scale thus easier to multimap change

**Comment by jknpj** [comment]
@Kurfursten not really, you can be an absolute shitcoder and just change their parent_type to avoid mapping changes.

**Comment by ShiftyRail** [comment]
Or just CTRL+H the .dmm files.

**Comment by Kurfursten** [comment]
I've done the find and replace dmm before and got mapping conflicts, never again on a major PR. This killed rigsuit PR, don't forget.

Also a thought: making this for use with a phone that's staying stationary (like red phones) would probably desire a more robust missed call system where it stores a list so you can check missed calls later having not had the phone on your person at the time. And ring more loudly maybe, showing the message to more than just its holder. I would also want to do different area code for landline.

File: code/game/objects/items/trader.dm [hunk 4]
```diff
@@ -1038,4 +1038,221 @@
...
+	connection = null
+	..()
+
+/obj/item/device/videophone/process()
+	var/mob/user = get_holder_of_type(src,/mob)
+	if(user)
+		to_chat(user,"<span class='warning'>[bicon(src)] *ring ring*</span>")
+	counter--
+	if(counter<=0)
+		to_chat(user,"<span class='info'>[bicon(src)] Missed call from #[connection.number]</span>")
```
Comment (Kurfursten) [inline code/game/objects/items/trader.dm]: needs sanity for user

File: code/game/objects/items/trader.dm [hunk 5]
```diff
@@ -1038,4 +1038,221 @@
...
+/obj/item/device/videophone/Move(NewLoc, Dir = 0, step_x = 0, step_y = 0, glide_size_override = 0)
+	..()
+	refresh_view()
+
+/obj/item/device/videophone/proc/refresh_view()
+	if(video_display)
+		video_display.vis_contents.Cut()
+		video_display.vis_contents = block_radius(src, PHONE_VIEW)
+
+/obj/item/device/videophone/dropped(mob/user)
```
Comment (Kurfursten) [inline code/game/objects/items/trader.dm:1258]: add key on pickup too

**Comment by DamianX** [comment]
Reopen if you come back to it.

**Pull request closed without merging**